apps/total-typescript/src/trpc/routers/module-progress.ts

```diff
@@ -0,0 +1,104 @@
+import {Section} from '@skillrecordings/skill-lesson/schemas/section'
+import {Lesson} from '@skillrecordings/skill-lesson/schemas/lesson'
+import {prisma} from '@skillrecordings/database'
+import {getToken} from 'next-auth/jwt'
+import {ModuleProgressSchema} from '../../video/module-progress'
+
+export const moduleProgressRouter = router({
+  bySlug: publicProcedure
+    .input(
+      z.object({
+        slug: z.string().nullish(),
+      }),
+    )
+    .query(async ({ctx, input}) => {
+      const token = await getToken({req: ctx.req})
+      if (!token || !input.slug) {
+        return null
+      }
+
+      const module = await getWorkshop(input.slug)
+
+      const allModuleLessons = module.sections.flatMap(
+        (section: Section) => section.lessons,
+      )
+      const lessonIds = allModuleLessons.map((lesson: Lesson) => lesson._id)
+
+      const moduleLessonProgress = await prisma.lessonProgress.findMany({
+        where: {
+          lessonId: {in: lessonIds},
+          userId: token.id as string,
+        },
+      })
+
+      const moduleProgressLessons = allModuleLessons.map((lesson: Lesson) => {
+        return {
+          id: lesson._id,
+          slug: lesson.slug,
+          lessonCompleted: Boolean(
+            moduleLessonProgress.find(
+              (progress) => progress.lessonId === lesson._id,
+            ),
+          ),
+        }
+      })
+
+      const moduleProgressSections = module.sections.map((section: Section) => {
+        const sectionProgressLessons =
+          section.lessons?.map((lesson: Lesson) => {
+            return {
+              id: lesson._id,
+              slug: lesson.slug,
+              lessonCompleted: Boolean(
+                moduleLessonProgress.find(
+                  (progress) => progress.lessonId === lesson._id,
+                ),
+              ),
+            }
+          }) || []
+        return {
+          id: section._id,
+          slug: section.slug,
+          sectionCompleted: sectionProgressLessons.every(
+            (lesson) => lesson.lessonCompleted,
+          ),
+          percentComplete: Math.round(
+            (sectionProgressLessons.filter((lesson) => lesson.lessonCompleted)
+              .length /
+              sectionProgressLessons.length) *
+              100,
+          ),
+          completedLessonCount: sectionProgressLessons.filter(
+            (lesson) => lesson.lessonCompleted,
+          ).length,
+          lessonCount: sectionProgressLessons.length,
+          lessons: sectionProgressLessons,
+        }
+      })
+
+      return ModuleProgressSchema.parse({
+        moduleId: module._id,
+        nextLesson:
+          moduleProgressLessons.find(
+            (lesson: {lessonCompleted: boolean}) => !lesson.lessonCompleted,
+          ) || null,
+        nextSection:
+          moduleProgressSections.find(
+            (section: {sectionCompleted: boolean}) => !section.sectionCompleted,
+          ) || null,
```

Comment:
does this need to be more complex?

Reply:
I can't think of any reason why that would be the case